rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,443 @@
+
+==== FeeConfig
+
+Each pool in Pablo defines a fee percentage to be charged for each trade. Except for LBPs other pools also define an owner fee that is a percentage out of the main trading fee. The `FeeConfig` is a new abstraction over all fees that could be charged on a pool to allow for extension. At this time a 100% of the owner fee should be defined as a new field `protocol_fee`.
+
+[plantuml,images/pablo-fee-config,png]
+----
+class FeeConfig {
+    +fee_rate: Permill
+    +owner_fee_rate: Permill,
+    +protocol_fee_rate: Permill,
+}
+----
+
+*Existing code must be modified to use this data structure*.
+
+Given this,
+----
+fee = // calculation depends on the pool type: based on the fee_rate
+owner_fee = fee * owner_fee_rate;
+protocol_fee = owner_fee * protocol_fee_rate;
+----
+For all pools launched at the Picasso launch following values would be set for these configs
+----
+owner_fee_rate = 20%
+protocol_fee_rate = 100% // all owner fees goes to composable to be distributed as rewards
+----
+
+==== LP Trading Fee Distribution
+
+Pablo needs to send the accumulated fees from trading to an account to be distributed asynchronously. It is natural to see this distribution working in a similar way to how staking rewards pallet distributes fNFT rewards. The idea behind this section is to reuse staking rewards pallet logic in distributing trading fees earnings to LPs.
+
+===== LP fNFT Configuration
+
+[plantuml,images/pablo-LP-fNFT-config,png]
+----
+start
+->after pablo pool creation,
+LP token: LPT,
+base token: BASE,
+quote token: QUOTE;
+:staking_rewards::configure(
+...
+asset: LPT,
+configuration: (
+    duration_presets: NO_END,
+    reward_assets: [BASE, QUOTE]),
+    early_unstake_penaly: 0
+);
+stop
+----
+
+The interface for the configurations does not currently exist on staking_rewards pallet, hence it has to be implemented.
+
+===== LP fNFT Creation When Receiving LP Tokens for Adding Liquidity
+
+[plantuml,images/pablo-fNFT-add-liquidity,png]
+----
+start
+->amount=x, token=LPT // LP tokens;
+:staking_rewards::stake(
+asset=LPT,
+from=lp_account_id,
+amount=x,
+duration=NO_END,
+keep_alive=false
+);
+stop
+----
+
+===== Trading Fee Distribution
+
+This reuses existing fNFT logic to distribute fees accrued for LPs.
+
+[plantuml,images/pablo-fNFT-fee-distro,png]
+----
+start
+->after Pablo swap,
+fee=x,
+fee token=QUOTE,
+LP token = LPT,
+from = transaction_origin;
+:StakingReward::transfer_reward(
+asset=LPT,
+reward_asset: QUOTE
+from=transaction_origin,
+amount=x,
+keep_alive=false
+);
+stop
+----
+
+==== PBLO Staker Trading Fee Distribution
+
+This is the reward a `PBLO` staker receives from the trading fees of Pablo pools. It is equal to the protocol fee charged on Pablo pools. This can be accomplished by calling the already exisiting `StakingReward.transfer_reward` interface as follows. According to product there is also a need to convert whatever the fee asset in to PBLO to create an demand/additional value for PBLO.
```

Comment:
```suggestion
This is the reward a `PBLO` staker receives from the trading fees of Pablo pools. It is equal to the protocol fee charged on Pablo pools. This can be accomplished by calling the already exisiting `StakingReward.transfer_reward` interface as follows. According to product there is also a need to convert whatever the fee asset in to PBLO to create a demand/additional value for PBLO.
```